frame/vamm/src/tests/swap_asset.rs

```diff
@@ -0,0 +1,477 @@
+		// `Overflow`, `Underflow`, etc.
+		swap_config.input_amount = 0;
+
+		swap_config.output_amount_limit = limit;
+		swap_config.vamm_id = VammId::zero();
+
+		ExtBuilder {
+			vamm_count: 1,
+			vamms: vec![(0, vamm_state)]
+		}.build().execute_with(|| {
+			assert_noop!(
+				TestPallet::swap(&swap_config),
+				Error::<MockRuntime>::SwappedAmountLessThanMinimumLimit
+			);
+		})
+	}
+}
+
+proptest! {
+	#![proptest_config(ProptestConfig::with_cases(RUN_CASES))]
+	#[test]
+	fn swap_asset_suceeds_emitting_event(
+		mut vamm_state in get_vamm_state(Default::default()),
+		mut swap_config in get_swap_config(Default::default()),
+		base in balance_range_low(),
+		quote in balance_range_low(),
+		peg in balance_range_low(),
+	) {
+		// Ensure vamm is open before start operation to swap assets.
+		vamm_state.closed = None;
+
+		// Ensure values won't overflow when computing swap.
+		vamm_state.base_asset_reserves = base;
+		vamm_state.quote_asset_reserves = quote;
+		vamm_state.peg_multiplier = peg;
+
+		// Disable output limit check
+		swap_config.output_amount_limit = Balance::zero();
+
+		swap_config.vamm_id = VammId::zero();
+
+		ExtBuilder {
+			vamm_count: 1,
+			vamms: vec![(0, vamm_state)]
+		}.build().execute_with(|| {
+			// For event emission
+			run_to_block(1);
+
+			let swap = TestPallet::swap(&swap_config);
+
+			// TODO(Cardosaum): Remove this `if let Ok()...` block and migrate
+			// to use `assert_ok!(swap)`. this is just a hack until all swap
+			// operations are fixed.
+			if let Ok(swap) = swap {
+				System::assert_last_event(
+					Event::Swapped {
+						vamm_id: swap_config.vamm_id,
+						input_amount: swap_config.input_amount,
+						output_amount: swap,
+						input_asset_type: swap_config.asset,
+						direction: swap_config.direction,
+					}.into()
+				);
+			}
+		})
+	}
+}
+
+#[test]
+fn swap_add_base() {
+	let swap_config = SwapConfig {
+		vamm_id: 0,
+		asset: AssetType::Base,
+		input_amount: 1_000_000_000_000,
+		direction: Direction::Add,
+		output_amount_limit: 0,
+	};
+
+	let base_u256 = U256::from(2_u128) * U256::exp10(12);
+	let quote_u256 = U256::from(50_u128) * U256::exp10(12);
+	let invariant = base_u256 * quote_u256;
+
+	ExtBuilder {
+		vamm_count: 1,
+		vamms: vec![
+			// Initial state
+			(
+				0,
+				VammState {
+					base_asset_reserves: base_u256.as_u128(),
+					quote_asset_reserves: quote_u256.as_u128(),
+					peg_multiplier: 1,
+					invariant,
+					closed: None,
+				},
+			),
+			// Expected final state
+			(
+				1,
+				VammState {
+					base_asset_reserves: 3_000_000_000_000,
+					quote_asset_reserves: 33_333_333_333_333,
+					peg_multiplier: 1,
+					invariant,
+					closed: None,
+				},
+			),
+		],
+	}
+	.build()
+	.execute_with(|| {
+		let swap = TestPallet::swap(&swap_config);
+		let vamm_after_swap = VammMap::<MockRuntime>::get(0);
+		let vamm_expected = VammMap::<MockRuntime>::get(1);
+
+		assert_eq!(swap, Ok(16_666_666_666_667));
+		assert_eq!(vamm_after_swap, vamm_expected);
```

Comment:
Maybe nit-picky, but did you need to push the expected state into storage to compare it later? I think you can just use it inside the closure (I use configs created outside the closure for `execute_with` in some of my tests).

Also remember that you may use `assert_ok!(swap, 16_666_666_666_667);`

Reply:
Yeah, technically I didn't need to insert it into storage. I'll change it, thanks for the input! :)

Regarding the `assert_ok`, I sometimes forget it exists :see_no_evil: haha